File: app/src/main/java/com/zhang/readme/view/base/BaseViewHolder.java
package com.zhang.readme.view.base;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import butterknife.ButterKnife;

/**
 * Created by zhang on 2017/2/22.
 *
 * @author zhang
 */

public abstract class BaseViewHolder<M> extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {

    public interface OnItemClickListener {
        void onItemClick(View view, int position);
        void onItemLongClick(View view, int position);
    }

    private OnItemClickListener mListener;

    public BaseViewHolder(@NonNull View itemView) {
        this(itemView, null, false);
    }

    public BaseViewHolder(@NonNull View itemView, @NonNull OnItemClickListener listener) {
        this(itemView, listener, true);
    }

    private BaseViewHolder(@NonNull View itemView, @Nullable OnItemClickListener listener, boolean clickable) {
        super(itemView);
        ButterKnife.bind(this, itemView);
        if (clickable) {
            mListener = listener;
            itemView.setOnClickListener(this);
            itemView.setOnLongClickListener(this);
        }
    }

    public abstract void bind(@NonNull M m);

    @Override
    public void onClick(View view) {
        int position = getAdapterPosition();
        if (position != RecyclerView.NO_POSITION) mListener.onItemClick(view, position);
    }

    @Override
    public boolean onLongClick(View view) {
        int position = getAdapterPosition();
        if (position != RecyclerView.NO_POSITION) mListener.onItemLongClick(view, position);
        return true;
    }

}

File: app/src/main/java/cn/devifish/readme/view/base/BaseViewHolder.kt
package cn.devifish.readme.view.base

import android.support.v7.widget.RecyclerView
import android.view.View

/**
 * Created by zhang on 2017/2/22.
 * 基类
 * @author zhang
 */

abstract class BaseViewHolder<M> constructor(itemView: View, listener: BaseViewHolder.OnItemClickListener? = null) : RecyclerView.ViewHolder(itemView), View.OnClickListener, View.OnLongClickListener {

    interface OnItemClickListener {
        fun onItemClick(view: View, position: Int)
        fun onItemLongClick(view: View, position: Int)
    }

    private var listener: OnItemClickListener? = null
        set(value) {
            field = value
            itemView.setOnClickListener(this)
            itemView.setOnLongClickListener(this)
        }

    init {
        if (listener != null) {
            this.listener = listener
        }
    }

    abstract fun bind(m: M)

    override fun onClick(view: View) {
        val position = adapterPosition
        if (position != RecyclerView.NO_POSITION) listener!!.onItemClick(view, position)
    }

    override fun onLongClick(view: View): Boolean {
        val position = adapterPosition
        if (position != RecyclerView.NO_POSITION) listener!!.onItemLongClick(view, position)
        return true
    }

}
